test/json.d

```diff
@@ -0,0 +1,308 @@
+        }
+        if (*next == '{')
+        {
+            next++;
+            return JsonValue(parseObject());
+        }
+        if (*next == '"')
+        {
+            next++;
+            return JsonValue(parseString());
+        }
+        if (next.startsWith("null"))
+        {
+            next += 4;
+            return JsonValue.null_;
+        }
+        if (next.startsWith("true"))
+        {
+            next += 4;
+            return JsonValue(true);
+        }
+        if (next.startsWith("false"))
+        {
+            next += 5;
+            return JsonValue(false);
+        }
+        if (*next == '-' || (*next >= '0' && *next <= '9'))
+            return JsonValue(parseNumber());
+
+        // todo: handle number/null/true/false
+        if (*next == '\0')
+            throw parseError(line(next), "expected value but reached EOF");
+
+        throw parseError(line(next), "expected value but got '%s' (0x%x)",
+            *next, cast(ubyte)*next);
+    }
+    JsonValue[] parseArray()
+    {
+        JsonValue[] array;
+        skipWhitespace();
+        if (*next != ']')
+        {
+            for (;;)
+            {
+                array ~= parseValue();
+                skipWhitespace();
+                if (*next == ']')
+                    break;
+                if (*next != ',')
+                    throw parseError(line(next), "expected comma ',' but got '%s' (0x%x)", *next, cast(ubyte)*next);
+                next++;
+                skipWhitespace();
+           }
+        }
+        next++;
+        return array;
+    }
+    JsonObject parseObject()
+    {
+        JsonObject object;
+        skipWhitespace();
+        if (*next != '}')
+        {
+            for(;;)
+            {
+                if (*next != '"')
+                    throw parseError(line(next), "expected object property but got '%s' (0x%x)",
+                        *next, cast(ubyte)*next);
+                next++;
+                auto name = parseString();
+                skipWhitespace();
+                if (*next != ':')
+                    throw parseError(line(next), "all property names must be followed by a colon ':'");
+                next++;
+                object.members ~= JsonObjectMember(name, parseValue);
+                skipWhitespace();
+                if (*next == '}')
+                    break;
+                if (*next != ',')
+                    throw parseError(line(next), "expected comma ',' but got '%s' (0x%x)", *next, cast(ubyte)*next);
+                next++;
+                skipWhitespace();
+            }
+        }
+        next++;
+        return object;
+    }
+    string parseString()
+    {
+        auto start = next;
+        for(;; next++)
+        {
+            if(*next == '"')
+            {
+               auto str = start[0 .. next - start];
+               next++;
+               return str;
+            }
+        }
+    }
+    BigInt parseNumber()
```

Comment:
Json only supports floating point numbers:

> A number is a sequence of decimal digits with no superfluous leading zero. It may have a preceding minus
sign (U+002D). It may have a fractional part prefixed by a decimal point (U+002E). It may have an exponent,
prefixed by e (U+0065) or E (U+0045) and optionally + (U+002B) or – (U+002D). The digits are the code
points U+0030 through U+0039.

http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf